Add Alignment Records for GEM Geometry

watson-ij opened the pull request:
#### PR description:

This PR adds alignment records for the GEM, and adds code to the GEMGeometryBuilder to allow the alignment to be applied. The code was written by following that used for the CSC.

#### PR validation:

A test of the alignment showed that the GEM geometry has left-handed local coordinate systems, which are incompatible with the standard Alignment records, which store the detector rotations as Euler angles from the global coordinate system. I removed the explicit axis flip which was responsible for this, and reran an upgrade sample and checked that the GEM rechits and segments were still being built properly.

By default, the alignment is off (an applyAlignment parameter is introduced in the GEM geometry builder), to avoid picking up currently non-existing alignment records. This PR is a first step to allow alignment to be applied to GEM, and while we plan to build on this by writing GEM alignment code, for now the alignment will not be used. Therefore, the PR is expected to produce no changes in output.

@jshlee @kpedro88

watson-ij committed:
Add GEM alignment records and use in GEMGeometryESModule (behind flag for now)

watson-ij committed:
Register GEM alignment record plugins in CondCore/AlignmentPlugins

watson-ij committed:
In GEMGeometryBuilder, always apply inversion on newY to rolls

The only change this will do is to flip local z axis on even GEM
chamber eta partitions. Without this, the rotation matrix is actually
a reflection and can't be represented by euler angles (as required for
e.g. Alignment)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26320/9023

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @watson-ij (Ian J. Watson) for master.

It involves the following packages:

CondCore/AlignmentPlugins
CondFormats/AlignmentRecord
Geometry/GEMGeometry
Geometry/GEMGeometryBuilder
Geometry/Records


@civanch, @Dr15Jones, @cvuosalo, @ianna, @tocheng, @cmsbuild, @franzoni, @mdhildreth, @kpedro88, @ggovi, @pohsun can you please review it and eventually sign? Thanks.
@dildick, @mmusich, @seemasharmafnal, @tocheng this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c8dba6a5a90a5f24589c3b3b92e5ad12cfeef9cc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/33903/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/33903/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 136 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3139747
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3139549
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

kpedro88 commented on Geometry/GEMGeometryBuilder/plugins/GEMGeometryESModule.cc:
```diff
@@ -20,13 +20,23 @@
+  : alignmentsLabel(p.getParameter<std::string>("alignmentsLabel"))
 {
   useDDD = p.getParameter<bool>("useDDD");
+  applyAlignment = p.getParameter<bool>("applyAlignment");
```
all of the variables could be initialized in the constructor initialization list

kpedro88 commented:
@watson-ij there are some differences in muon reco quantities in Phase 2 workflows. It seems likely that these differences are caused by the change in the coordinate system. Can you check it carefully?

watson-ij committed:
Merge branch 'master' into gem-align-master

Including MuonGeometryRecord changes

watson-ij committed:
Move variable setup into initializer list

cmsbuild commented:
The code-checks are being triggered in jenkins.

watson-ij commented:
@kpedro88, the differences are mostly from a single workflow, 20034.0, so I ran this on a nightly from 04-01, and with and without my changes it gave the reference distribution. I then tried a 04-07 nightly and with or without my changes, it produced the "incorrect" result. I'm not sure if its possible another commit caused changes from the reference without updating the reference? I also fixed a merge conflict with a recent commit, and implemented the changes you suggested in the constructor. Could we try rerunning the comparison checks?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26320/9159

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #26320 was updated. @civanch, @Dr15Jones, @cvuosalo, @ianna, @tocheng, @cmsbuild, @franzoni, @mdhildreth, @kpedro88, @ggovi, @pohsun can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/34073/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 134 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140297
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

cvuosalo commented:
Geometry is ready to approve if the comparison test differences can be explained.

watson-ij commented:
Okay, I can see the differences in the workflow now when I take the last nightly with and without my changes. I'll see if I can debug whats going on.

pohsun commented:
No significant flaw in current commit. Alca is pending for updates.

kpedro88 commented:
@watson-ij what is the status of this PR?

watson-ij commented:
@kedpro88, I've been drilling into the code to see the changes, and
think I know the reason, but it would be good for a tracking expert to
cross-check.

The plot differences are mostly from a single standalone displaced
muon missing after my PR, but another GEMMuon having the standalone
flag on that previously wasn't.

After tracing through several systems, I found that the PR changes the
results of the kalman update step in the even chambers which have
their z-flipped (in particular, I checked the muon reco systems, such
as the fast detector hit finders, and none of these is changed by the
PR).

For example, first muon passing through GEM in event 10 has these
local parameters after the Kalman update (before the update they are
identical):

Before PR:
local parameters (q/p,v',w',v,w)
     0.355115   -0.0415613     0.241321      8.87442     0.250691

After PR:
local parameters (q/p,v',w',v,w)
    -0.430425    0.0415613    -0.241321      8.87442     0.250691

The v'=dv/du (u being along the local z), w' sign flip makes sense,
since we flipped the z-axis, so only the q/p is different. Looking at
what passes through the KFUpdator step, the measurement parameters are
all the same, the difference is coming from the
BasicTrajectoryState::createLocalErrorFromCurvilinearError() which
uses JacobianCurvilinearToLocal curv2Loc to get the jacobian and then
uses it to translate the curvilinear covariance matrix of the track to
the local parameters.

The paper describing this transformation is here:
https://doi.org/10.1016/j.nima.2006.07.032 and in it, they state
several times that they are defining axes to be right-handed, so I
think that using left-handed coordinates is wrong and only
right-handed coordinates should be used, so this PR (which is changing
all local frame to be right-handed) would fix that issue.

I was trying to follow through the derivation in this particular case
to confirm this, but getting bogged down in the math. So, it would be
good for a tracking expert to weigh in if my understanding here is
correct, and any additional suggestions would be welcome.

kpedro88 commented:
@watson-ij thanks for investigating this. I'll tag some tracking experts here, since I am not one: @VinInn @makortel @JanFSchulte @mtosi, can you check if the above explanation makes sense?

kpedro88 commented:
ping @ tracking experts (@VinInn @makortel @JanFSchulte @mtosi)

kpedro88 commented:
ping @ tracking experts (@VinInn @makortel @JanFSchulte @mtosi)

kpedro88 commented:
+upgrade
seeing no objection from tracking experts, the proposed explanation for observed changes looks reasonable to me

pohsun commented:
+1

pohsun commented:
+1

ggovi commented:
+1

fabiocos commented:
please test

fabiocos commented:
@cvuosalo could you please come back to this PR? Your signature is the last missing

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3691500bdd08b5424c3d630dac60515a768daf48
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8044f7/249/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/35d2893ef6d1dfc9bf110cb951516b00d926eebf
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/249/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/249/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
+1
Tested at: 3691500bdd08b5424c3d630dac60515a768daf48
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8044f7/245/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/35d2893ef6d1dfc9bf110cb951516b00d926eebf
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/245/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26320/245/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8044f7/249/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 267 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212004
 - DQMHistoTests: Total failures: 474
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211196
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cvuosalo commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
